README.md

#### General

[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active) [![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.1095473.svg)](https://doi.org/10.5281/zenodo.1095473)

| Resource:     | CRAN                                                                                                                                                                     | Travis CI                                                                                                                                               | Appveyor                                                                                                                                                                      |
|---------------|--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| *Platforms:*  | *Multiple*                                                                                                                                                               | *Linux & macOS*                                                                                                                                         | *Windows*                                                                                                                                                                     |
| R CMD check   | <a href="https://cran.r-project.org/web/checks/check_results_oddsratio.html"><img border="0" src="http://www.r-pkg.org/badges/version/oddsratio" alt="CRAN version"></a> | <a href="https://travis-ci.org/pat-s/oddsratio"><img src="https://travis-ci.org/pat-s/oddsratio.svg?branch=dev" alt="Build status"></a>                 | <a href="https://ci.appveyor.com/project/pat-s/oddsratio"><img src="https://ci.appveyor.com/api/projects/status/p72njexjxqa9ko96/branch/dev?svg=true" alt="Build status"></a> |
| Test coverage |                                                                                                                                                                          | <a href="https://codecov.io/gh/pat-s/oddsratio"><img src="https://codecov.io/gh/pat-s/oddsratio/branch/dev/graph/badge.svg" alt="Coverage Status"/></a> |                                                                                                                                                                               |

#### CRAN

[![CRAN\_Status\_Badge](http://www.r-pkg.org/badges/version/oddsratio)](https://cran.r-project.org/package=oddsratio) [![Downloads](https://cranlogs.r-pkg.org/badges/oddsratio?color=brightgreen)](https://www.r-pkg.org/pkg/oddsratio) ![](https://cranlogs.r-pkg.org/badges/grand-total/oddsratio)

Functions for calculation and plotting of odds ratios of Generalized Additive (Mixed) Models and Generalized Linear (Mixed) Models with a binomial response variable (i.e. logistic regression models).

Installation
------------

Install from CRAN:

``` r
install.packages("oddsratio")
```

Get the development version from Github:

``` r
remotes::install_github("pat-s/oddsratio")
```

Examples
--------

### GLM

Odds ratio calculation of predictors `gre` & `gpa` of a fitted model `fit_glm` with increment steps of 380 and 5, respectively.
For factor variables (here: `rank` with 4 levels), automatically all odds ratios corresponding to the base level (here: `rank1`) are returned including their respective confident intervals. The default level is 95%. However, other levels can be specified with the param `CI`. Data source: <http://www.ats.ucla.edu/stat/r/dae/logit.htm>

``` r
pacman::p_load(oddsratio, mgcv)
df <- data_glm
df$rank <- factor(df$rank)
fit_glm <- glm(admit ~ gre + gpa + rank, data = df, family = "binomial")

or_glm(data = df, model = fit_glm, 
       incr = list(gre = 380, gpa = 5, CI = 0.95))
```

### GAM

For GAMs, the calculation of odds ratio is different. Due to its non-linear definition, odds ratios do only apply to specific value changes and are not constant throughout the whole value range of the predictor as for GLMs. Hence, odds ratios of GAMs can only be computed for one predictor at a time by holding all other predictors at a fixed value while changing the value of the specific predictor. Confident intervals are currently fixed to the 95% level for GAMs. Data source: `?mgcv::predict.gam()`

Here, the usage of `or_gam()` is shown by calculating odds ratios of pred `x2` for a 20% steps across the whole value range of the predictor.

``` r
set.seed(1234)
n <- 200
sig <- 2
df <- gamSim(1, n = n,scale = sig, verbose = FALSE)
df$x4 <- as.factor(c(rep("A", 50), rep("B", 50), rep("C", 50), rep("D", 50)))
fit_gam <- mgcv::gam(y ~ s(x0) + s(I(x1^2)) + s(x2) + offset(x3) + x4, data = df)

or_gam(data = df, model = fit_gam, pred = "x2",
       percentage = 20, slice = TRUE)
```

If you want to compute a single odds ratio for specific values, simply set param `slice = FALSE`:

``` r
or_gam(data = df, model = fit_gam,
       pred = "x2", values = c(0.099, 0.198))
```

Plotting of GAM smooths is also supported:

``` r
plot_gam(fit_gam, pred = "x2", title = "Predictor 'x2'")
```

<p align="center">
<img src="https://raw.githubusercontent.com/pat-s/oddsratio/dev/man/figures/plot_gam.png" width="80%"/>
</p>
Insert the calculated odds ratios into the smoothing function:

``` r
plot_object <- plot_gam(fit_gam, pred = "x2", title = "Predictor 'x2'")
or_object <- or_gam(data = df, model = fit_gam, 
                    pred = "x2", values = c(0.099, 0.198))

plot <- insert_or(plot_object, or_object, or_yloc = 3,
                  values_xloc = 0.04, line_size = 0.5, 
                  line_type = "dotdash", values_yloc = 0.5,
                  arrow_col = "red")
plot
```

<p align="center">
<img src="https://raw.githubusercontent.com/pat-s/oddsratio/dev/man/figures/plot_gam2.png" width="80%"/>
</p>
Insert multiple odds ratios into one smooth:

``` r
or_object2 <- or_gam(data = df, model = fit_gam, pred = "x2", 
                     values = c(0.4, 0.6))

insert_or(plot, or_object2, or_yloc = 2.1, values_yloc = 2,
          line_col = "green4", text_col = "black",
          rect_col = "green4", rect_alpha = 0.2,
          line_alpha = 1, line_type = "dashed",
          arrow_xloc_r = 0.01, arrow_xloc_l = -0.01,
          arrow_length = 0.01, rect = T)   
```

<p align="center">
<img src="https://raw.githubusercontent.com/pat-s/oddsratio/dev/man/figures/plot_gam3.png" width="80%"/>
</p>


NEWS
oddsratio 1.0.3 (June 19 2018)
--------------------------------------------------------------------------------

 * update functions to work with ggplot2 v3.0.0

oddsratio 1.0.2 (December 08 2017)
--------------------------------------------------------------------------------

Minor:
  * Add CITATION file

oddsratio 1.0.0 (June 12 2017)
--------------------------------------------------------------------------------

Major:
  * rename functions (snake_case)

oddsratio 0.3.1 (Nov 9 2016)
--------------------------------------------------------------------------------

 * update functions to work with ggplot2 v2.2.0
 * add data and enable lazy loading in examples

oddsratio 0.3.0 (Oct 27 2016)
--------------------------------------------------------------------------------

# New functions:
--------------------------------------------------------------------------------
 * `plot_smooth.gam()`: Lets you plot smoothing functions of GAM(M)s using `ggplot2`.
 * `add.oddsratio.into.plot()`: Add odds ratios into plot of GAM(M) smoothing function.

# Function updates:
--------------------------------------------------------------------------------
 * `calc.oddsratio.glm`, `calc.oddsratio.gam`: Add odds ratio confident interval calculation 
 * For GLM models CI level can be specified manually.
 * Print 'CI' warning if model is of type `glmmPQL`

oddsratio 0.2.0 (Oct 12 2016)
--------------------------------------------------------------------------------

 * Remove param `quietly`
 * return data.frame in any case
 * update DESCRIPTION

oddsratio 0.1.0 (Oct 11 2016)
--------------------------------------------------------------------------------

 * Initial release attempt to CRAN


R/insert_or.R
#' @name insert_or
#' @title Insert odds ratios of GAM(M)s into smoothing function
#'
#' @description This function inserts calculated odds ratios of GAM(M)s into
#' a plot of a GAM(M) smoothing function.
#'
#' @import ggplot2
#' @importFrom cowplot background_grid
#'
#' @param plot_object A `ggplot` object from [plot_gam].
#'
#' @param or_object A returned data.frame from [or_gam].
#'
#' @param values Logical. Whether to print predictor value information nearby
#' the inserted vertical lines. Default to `TRUE`.
#'
#' @param rect Logical. Whether to print a shaded rectangle between the
#' vertical lines.
#'
#' @param arrow Logical. Whether to print arrows above the inserted values.
#' Default to `TRUE`.
#'
#' @param line_col,line_alpha,line_type,line_size Aesthetics of vertical lines.
#'
#' @param text_col,text_alpha,text_size Aesthetics of inserted values.
#'
#' @param rect_col,rect_alpha Aesthetics of shaded rectangle.
#'
#' @param or_yloc,values_yloc Numeric. Specifies y-location of inserted
#' odds ratio / values.
#' Relative to plotted y-axis range. A positive/negative value will place the
#' the text higher/lower.
#'
#' @param values_xloc Numeric. X-axis location/shift of values relative to
#' their vertical line.
#' Default to 2\% of x-axis range.
#'
#' @param arrow_xloc_r,arrow_xloc_l,arrow_yloc,arrow_length,arrow_col Numeric.
#' Axis placement options of inserted arrows.
#' Relative to respective axis ranges.
#'
#' @details The idea behind this function is to add calculated odds ratio of
#' fitted GAM models ([or_gam]) into a plot
#' showing the smooth function ([plot_gam]) of the chosen
#' predictor for which the odds ratio was calculated for. Multiple insertions
#' can be made by iteratively calling the function (see examples).
#'
#' @details Right now the function does only accept results of
#' [or_gam] with `slice = FALSE`.
#' If you want to insert multiple odds ratio you have to do it iteratively.
#'
#' @return Returns a `ggplot` plotting object
#'
#' @seealso [plot_gam]
#' @seealso [or_gam]
#'
#' @author Patrick Schratz <patrick.schratz@gmail.com>
#'
#' @examples
#' # load data (Source: ?mgcv::gam) and fit model
#' library(mgcv)
#' fit_gam <- gam(y ~ s(x0) + s(I(x1^2)) + s(x2) +
#'                offset(x3) + x4, data = data_gam) # fit model
#'
#' # create input objects (plot + odds ratios)
#' library(oddsratio)
#' plot_object <- plot_gam(fit_gam, pred = "x2", title = "Predictor 'x2'")
#' or_object1 <- or_gam(data = data_gam, model = fit_gam,
#'                      pred = "x2", values = c(0.099, 0.198))
#'
#' # insert first odds ratios to plot
#' plot_object <- insert_or(plot_object, or_object1, or_yloc = 3,
#'                          values_xloc = 0.04, line_size = 0.5,
#'                          line_type = "dotdash", text_size = 6,
#'                          values_yloc = 0.5, arrow_col = "red")
#'
#' # calculate second odds ratio
#' or_object2 <- or_gam(data = data_gam, model = fit_gam, pred = "x2",
#'                      values = c(0.4, 0.6))
#'
#' # add or_object2 into plot
#' insert_or(plot_object, or_object2, or_yloc = 2.1, values_yloc = 2,
#'           line_col = "green4", text_col = "black",
#'           rect_col = "green4", rect_alpha = 0.2,
#'           line_alpha = 1, line_type = "dashed",
#'           arrow_xloc_r = 0.01, arrow_xloc_l = -0.01,
#'           arrow_length = 0.01, rect = TRUE)
#' @export

insert_or <- function(plot_object = NULL, or_object = NULL, line_col = "red", # nocov start # nolint
                      line_size = 1.2, line_type = "solid", line_alpha = 1,
                      text_alpha = 1, text_size = 4, text_col = "black",
                      rect_alpha = 0.5, rect_col = NULL,
                      rect = FALSE, arrow = TRUE, values = TRUE,
                      values_yloc = 0, values_xloc = NULL, or_yloc = 0,
                      arrow_length = NULL, arrow_yloc = NULL, arrow_col = NULL,
                      arrow_xloc_r = NULL, arrow_xloc_l = NULL) {

  plot_object <- plot_object +
    geom_vline(xintercept = or_object$value1, color = line_col,
               size = line_size, linetype = line_type, alpha = line_alpha) +
    geom_vline(xintercept = or_object$value2, color = line_col,
               size = line_size, linetype = line_type, alpha = line_alpha) +
    annotate("text", x = mean(c(or_object$value2, or_object$value1)),
             y = min(plot_object$data$se_lwr) + or_yloc,
             label = paste0("OR: \n", round(or_object$oddsratio, 2)),
             color = text_col, size = text_size)

  if (rect) {
    if (is.null(rect_col)) {
      rect_col <- text_col
    }

    # set drawing order to place rect behind smoothing fun
    plot_object$layers <- c(geom_rect(data = plot_object$data[1,], # avoids multiple rect drawings # nolint
                                      ymin = ggplot_build(plot_object)$layout$
                                        panel_params[[1]]$y.range[1],
                                      ymax = ggplot_build(plot_object)$layout$
                                        panel_params[[1]]$y.range[2],
                                      xmin = or_object$value1,
                                      xmax = or_object$value2,
                                      alpha = rect_alpha, fill = rect_col),
                            plot_object$layers)
  }

  if (values) {
    if (is.null(values_xloc)) {
      # calc x range for x-shift
      values_xloc <- 0.03
      values_xloc <- (max(plot_object$data$x) - min(plot_object$data$x)) *
        values_xloc
    }

    if (is.null(arrow_length)) {
      # calc arrow length from x axis range
      arrow_length <- (max(plot_object$data$x) - min(plot_object$data$x)) * 0.01
    }

    if (is.null(arrow_yloc)) {
      # calc arrow height from y axis range
      arrow_yloc <- (max(plot_object$data$y) - min(plot_object$data$y)) * 0.05
    }

    if (is.null(arrow_xloc_l)) {
      # calc arrow shift from x axis range
      arrow_xloc_l <- - (max(plot_object$data$y) - min(plot_object$data$y)) *
        0.002
    }

    if (is.null(arrow_xloc_r)) {
      # calc arrow shift from x axis range
      arrow_xloc_r <- (max(plot_object$data$y) - min(plot_object$data$y)) *
        0.002
    }

    plot_object <- plot_object +

      annotate("text", x = or_object$value1 - values_xloc,
               y = min(plot_object$data$se_lwr) + values_yloc,
               label = or_object$value1,
               color = text_col, alpha = text_alpha, size = text_size) +
      annotate("text", x = or_object$value2 + values_xloc,
               y = min(plot_object$data$se_lwr) + values_yloc,
               label = or_object$value2,
               color = text_col, alpha = text_alpha, size = text_size)

    if (arrow) {

      if (is.null(arrow_col)) {
        # calc arrow shift from x axis range
        arrow_col <- text_col
      }

      plot_object <- plot_object +

        # left arrow
        geom_segment(x = or_object$value1 - values_xloc + arrow_xloc_l,
                     xend = or_object$value1 - values_xloc + arrow_length,
                     y = min(plot_object$data$se_lwr) + values_yloc +
                       arrow_yloc,
                     yend = min(plot_object$data$se_lwr) + values_yloc +
                       arrow_yloc,
                     color = arrow_col,
                     arrow = arrow(length = unit(0.2, "cm"), type = "closed")) +
        # right arrow
        geom_segment(x = or_object$value2 + values_xloc + arrow_xloc_r,
                     xend = or_object$value2 + values_xloc - arrow_length,
                     y = min(plot_object$data$se_lwr) + values_yloc +
                       arrow_yloc,
                     yend = min(plot_object$data$se_lwr) + values_yloc +
                       arrow_yloc,
                     color = arrow_col,
                     arrow = arrow(length = unit(0.2, "cm"), type = "closed"))
    }
  }
  return(plot_object)
} # nocov end
